### PR TITLE
added LAPACK dsyevd and dsyevr for symEig

### PR DESCRIPTION
**Pull request by jbloom22**
-added symEigD and symEigR in stats on Breeze matrices
-added unit test and speed test

### CODE REVIEW

**Comment by jbloom22** [comment]
The Breeze method symEig calls to LAPACK dsyev, which performs poorly relative to dsyevd and dsyevr:

http://www.netlib.org/lapack/lawnspdf/lawn183.pdf

http://scicomp.stackexchange.com/questions/11827/flop-counts-for-lapack-symmetric-eigenvalue-routines-dsyev-dsyevd-dsyevx-and-d

This data is from the symEigSpeedTest in symEigDSuite on MacBook Pro:

[symEig6k.pdf](https://github.com/hail-is/hail/files/512569/symEig6k.pdf)

```
dim	svd	symEigD symEig symEigR
500	.092137	.051487	.234	.038
500	.088740	.050629	.217	.046
500	.093237	.047375	.229	.041
1000	.458839	.193334	1.659	.191
1000	.430039	.184285	1.469	.195
1000	.441143	.207065	1.464	.183
1500	1.399	.724547	4.810	.595
1500	1.407	.599062	4.777	.601
1500	1.421	.583506	5.236	.627
2000	3.272	1.479	10.942	1.386
2000	3.205	1.337	11.006	1.381
2000	3.473	1.354	10.933	1.366
2500	6.180	2.519	21.639	2.750
2500	6.217	2.718	21.772	2.758
2500	6.580	2.590	21.176	2.661
3000	10.169	4.117	51.154	4.716
3000	10.414	4.131	51.602	4.834
3000	10.709	4.219	46.711	4.794
3500	15.451	6.549	72.2	7.365
3500	15.353	7.058	75.9	7.194
3500	15.350	6.516	70.9	7.210
4000	20.584	9.111	112.6	10.725
4000	22.085	9.476	110.3	10.594
4000	21.920	9.461	108.2	11.062
4500	29.075	13.488	143.8	15.440
4500	30.305	13.402	140.2	15.338
4500	31.339	13.562	134.3	15.294
5000	43.908	17.818	196.0	21.286
5000	40.874	17.821	197.7	21.231
5000	41.582	18.088	198.9	21.357
5500	58.772	24.747	271.1	28.879
5500	60.6	23.844	269.8	28.130
5500	60.9	24.197	275.3	28.356
```

File: src/main/scala/org/broadinstitute/hail/stats/eigSymD.scala [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      def apply(X: DenseMatrix[Double]): DenseVector[Double] = {
+        doeigSymD(X, rightEigenvectors = false)._1
+      }
+    }
+
+  }
+
+  private def doeigSymD(X: Matrix[Double], rightEigenvectors: Boolean): (DenseVector[Double], Option[DenseMatrix[Double]]) = {
+    //requireNonEmptyMatrix(X)
+    //requireSymmetricMatrix(X)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/stats/eigSymD.scala]: If you leave these in commented out, make a short comment saying why they're there (e.g. assume X is non-empty, symmetrix.  Check too expensive, disabling by default.)  In general, I'm okay with (amortized) O(1) checks staying in at this point.

File: src/main/scala/org/broadinstitute/hail/stats/eigSymD.scala [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      def apply(X: DenseMatrix[Double]): DenseVector[Double] = {
+        doeigSymR(X, rightEigenvectors = false)._1
+      }
+    }
+
+  }
+
+  private def doeigSymR(X: Matrix[Double], rightEigenvectors: Boolean): (DenseVector[Double], Option[DenseMatrix[Double]]) = {
+    //requireNonEmptyMatrix(X)
+    //requireSymmetricMatrix(X)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/stats/eigSymD.scala]: Same.

File: src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        print("eigSymR: ")
+        printTime(computeEigSymR())
+        print("eigSymD: ")
+        printTime(computeEigSymD())
+        println()
+      }
+    }
+
+    // Do not run with standard tests
+    // timeSymEig()
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala]: Just leave off the `@Test`.

File: src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      assert(D_==(svdK.S(j), eigSymDK.eigenvalues(n - j - 1)))
+      for (i <- 0 until n) {
+        assert(D_==(math.abs(svdK.U(i, j)), math.abs(eigSymDK.eigenvectors(i, n - j - 1))))
+      }
+    }
+
+    // eigSymD = eigSym
+    for (j <- 0 until n) {
+      assert(D_==(eigSymK.eigenvalues(j), eigSymDK.eigenvalues(j)))
+      for (i <- 0 until n) {
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala]: You could define a matrix comparison helper function.

File: src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    val n = 5
+    val m = 10
+    val W = DenseMatrix.fill[Double](n, m)(rand.gaussian.draw())
+    val K = W * W.t
+
+    val svdW = svd(W)
+    val svdK = svd(K)
+    val eigSymK = eigSym(K)
+    val eigSymDK = eigSymD(K)
+    val eigSymRK = eigSymR(K)
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala]: Testing different implementations are great.  You might also test a small, concrete input.  If everything is systematically wrong you won't catch it.

File: src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala (hunk 3 above)
Comment (jbloom22) [inline src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala]: done

File: src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala (hunk 5 above)
Comment (jbloom22) [inline src/test/scala/org/broadinstitute/hail/stats/symEigDSuite.scala]: added one

**Comment by jbloom22** [comment]
back to you

File: src/test/scala/org/broadinstitute/hail/stats/eigSymDSuite.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    assert(D_==(math.abs(eigSymRK2.eigenvectors(1,1)), c))
+  }
+
+  def symEigSpeedTest() {
+    val seed = 0
+
+    val rand = new JDKRandomGenerator()
+    rand.setSeed(seed)
+
+    def timeSymEig() = {
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/stats/eigSymDSuite.scala]:
```
def f() <body>
f()
```

is the same as:

```
<body>
```

Also, be care of using `=` in a function definition without a type declaration.

**Pull request merged**